[Feat] /my-page/setting/manage-account 퍼블리싱 생성

yonghyeun opened the pull request:
# 관련 이슈 번호
close #199
# 설명

생각보다 작업 단위가 너무 컸네요 .. 

API 작업은 API 문서 나온 후 진행 하도록 합니다 
---
# 디자인 시스템 생성 

## Notice

![image](https://github.com/user-attachments/assets/24c96df7-0184-43a6-b231-ef75349d7686)
![image](https://github.com/user-attachments/assets/fcd54a45-ec98-4ea7-b4f7-bea6a22152e2)

피그마에 아주 조그맣게 있던 디자인 시스템이  탈퇴 시 나타나는 비밀번호 확인 란에 사용되더군요 

그래서 다음과 같이 디자인 시스템을 생성했습니다.

```tsx
interface NoticeProps {
  title?: React.ReactNode;
  children: React.ReactNode | [React.ReactNode, React.ReactNode];
  className?: string;
}
export const Notice = ({
  title,
  children,
  className = "rounded-2xl",
}: NoticeProps) => {
  return (
    <div
      className={`$flex flex-col gap-1 px-4 py-[.625rem] justify-center self-stretch rounded-2xl bg-grey-50 btn-3 text-grey-500 ${className}`}
    >
      {title}
      <div className="flex gap-1 body-3">{children}</div>
    </div>
  );
};
```

내부에 들어가는 아이콘들의 크기를 `w-20px h-20px` 로 맞춰줘야 하는데 어떻게 할지 감이 잘 안오더라고요 

{leadingIcon} 을 다른 태그로 감싸고 해당 태그의 크기를 정적으로 정해줘도 내부에 존재하는 아이콘의 크기가 변하지 않습니다 ㅠ_ㅠ 

그래서 그냥 사용 할 때는 이렇게 사용했습니다.

```tsx
        {/* 알림창 */}
        <Notice>
          <InfoIcon width={20} height={20} />
          <span>탈퇴 전 한번 더 비밀번호를 입력해 주세요</span>
        </Notice>
```

## ExitModal 

![image](https://github.com/user-attachments/assets/fcd4b27e-94f9-4275-9a8c-68fb983e5260)

폼에서 반복적으로 사용되는 `ExitModal` 을  디자인 시스템으로 만들어뒀습니다.

> TODO 리팩토링 단계에서 다른 모달에서 `ExitModal ` 을 사용하도록 해야합니다.

# 계정관리 퍼블리싱

![image](https://github.com/user-attachments/assets/29c31862-a116-47e8-aabe-889a143c87b8)

계정 관리에 해당하는 부분들을 퍼블리싱 했습니다.
 
```tsx
export const AccountManagementPage = () => {
  const navigate = useNavigate();
  const role = useAuthStore((state) => state.role);

  useEffect(() => {
    if (!role) {
      navigate(ROUTER_PATH.MY_PAGE);
    }
  }, [navigate, role]);

  if (!role) {
    return null;
  }

  return (
    <>
      <BackwardNavigationBar
        label={<h1 className="title-1 text-grey-700">계정 관리</h1>}
      />
      <section className="flex flex-col gap-4 px-4 py-4">
        <AccountEmail />
        <PasswordChangeButton />
        <DividerLine axis="row" />
        <AccountCancellation />
      </section>
    </>
  );
};
```
---
# PasswordChangeButton

해당 버튼은 비밀번호를 변경하는 버튼입니다. 해당 버튼을 클릭하면 `PasswordChangeModal` 이 띄워지게 됩니다.

## PasswordChangeModal

`PasswordChangeModal` 은 다음과 같이 생겼습니다.

![image](https://github.com/user-attachments/assets/c30e621f-e521-42c9-8be7-a86e851d2d67)

유효성 검사나 그런 부분들은 이전에 쏴님이 너무 잘 만들어주셔서 해당 부분들을 재사용했습니다.

> 하면서 느끼는건데 `validatePassword` 메소드도 `shared` 에 둬야 할 거 같습니다 
>
> 추후 대규모 리팩토링 때 우다다 해봅시다 

이번 작업에서 특징적인 부분은 어떻게 렌더링을 최소한으로 할 수 있을까  ? 였습니다.

그래서 고안한 방법이 최대한 `password , newPassword .. ` 등과 같이 자주 변경이 일어나는 상태를 구독하지 않는 비제어 컴포넌트 처럼 사용하는 거였습니다.

```tsx
const CurrentPasswordInput = () => {
  const isValidPassword = usePasswordChangeFormStore(
    (state) => state.isValidPassword,
  );
  const isEmptyCurrentPassword = usePasswordChangeFormStore(
    (state) => state.isEmptyCurrentPassword,
  );

  const setCurrentPassword = usePasswordChangeFormStore(
    (state) => state.setCurrentPassword,
  );
  const setIsValidPassword = usePasswordChangeFormStore(
    (state) => state.setIsValidPassword,
  );

  const setIsEmptyCurrentPassword = usePasswordChangeFormStore(
    (state) => state.setIsEmptyCurrentPassword,
  );

  const handleChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = target;
    setCurrentPassword(value);

    const isValidPassword = validatePassword(value);
    const isPasswordEmpty = value.length === 0;

    setIsValidPassword(isValidPassword);
    setIsEmptyCurrentPassword(isPasswordEmpty);
  };

  const statusText = isEmptyCurrentPassword
    ? "비밀번호를 입력해 주세요"
    : isValidPassword
      ? ""
      : "비밀번호 형식에 맞게 입력해 주세요";

  return (
    <PasswordInput
      id="current-password"
      label="현재 비밀번호"
      name="current-password"
      placeholder="현재 비밀번호를 입력해주세요"
      essential
      onChange={handleChange}
      statusText={statusText}
      isError={!isEmptyCurrentPassword && !isValidPassword}
    />
  );
};
```

해당 컴포넌트 양상을 보면 해당 컴포넌트는 `password` 를 스토어에서 변경 시키지만 구독하고 있지 않아 `password` 가 변경된다고 해서 매번 리렌더링이 일어나지 않습니다. 

리렌더링이 일어나는 경우는 `statusText` 가 변해야 하는 상황에서만 일어납니다 :) 

> ### PS
>
> 이전에 이야기 나눌 때 유효성 검사 로직을 스토어에 위임하고 컴포넌트에서 하지 않아보겠다 했는데, 그러면 알아보기 힘들 거 같아서 유효성 검사 로직을 컴포넌트에서 처리했습니다.
>
> `ConfirmPassword` 부분에서의 유효성 검사 로직이 그다지 복잡하지 않더군요 
```tsx
   // newPassword
    const { confirmPassword } = usePasswordChangeFormStore.getState();
    setIsValidConfirmPassword(value === confirmPassword);
   // ConfirmPassword
    const { newPassword } = usePasswordChangeFormStore.getState();
    setIsValidConfirmPassword(newPassword === value);
```

그냥 단순히 각 패스워드 영역에서 본인 값이 스토어에 저장된 다른 패스워드 입력 창의 값과 같은지만 확인하면 됐습니다.

---

# 탈퇴하기

![image](https://github.com/user-attachments/assets/cb456b44-5cfe-4b81-9022-5299d7a70701)
![image](https://github.com/user-attachments/assets/12bfa664-348f-4a6a-a719-cd323ba575fd)
![image](https://github.com/user-attachments/assets/1f19705f-9cc9-47e8-9d6e-98756a602739)

탈퇴하기는 `useModal` 을 계속 모달 내부에서 호출해서 3 depth 의 모달을 생성했습니다. 

만들면서 느낀게 `handle .. , on .. ` 이 부분이 서로 혼합되어서 쓰여서 뭐가 뭔지 구분하는게 쉽지 않네요 🔥🫠





# 첨부 파일 (Optional)

# 레퍼런스 (Optional)

yonghyeun committed:
feat[#199] : AccountEmail 컴포넌트 생성

yonghyeun committed:
feat[#199] : /my-page/setting/manage-account 페이지 생성

yonghyeun committed:
feat[#199] : usePasswordChangeFormStore 생성

yonghyeun committed:
feat[#199] : /my-page/manage-account 뷰 완성

yonghyeun committed:
feat[#199] : PasswordChangeModal Input 완성

yonghyeun committed:
feat[#199] : 자주 재사용되는 ExitModal 생성

yonghyeun committed:
feat[#199] : PasswordChangeModal 내부에서 ExitModal 호출 하도록 수정

yonghyeun committed:
feat[#199] : ExitModal 스토리북 추가

yonghyeun committed:
refactor[#199] : ExitModal import 경로 수정

yonghyeun committed:
feat[#199] : Notice 디자인 시스템 생성

yonghyeun committed:
feat[#199] : usePasswordChangeStore reset action 추가

yonghyeun committed:
feat[#199] : 계정 탈퇴 시 비밀번호 확인 모달 생성

yonghyeun committed:
feat[#199] : 계정 탈퇴 풀페이지 모달 생성

yonghyeun committed:
feat[#199] : /my-page/manage-account 권한에 따른 라우팅 처리

aken-you commented on src/shared/ui/notice/Notice.tsx:
```diff
@@ -0,0 +1,19 @@
+interface NoticeProps {
+  title?: React.ReactNode;
```
title 타입이 ReactNode인 이유가 있을까요?!

yonghyeun replied:
말 그대로 타이틀에 어떤 값이든 들어 올 수 있어서요 

string 이랑 JSX.Element 로만 타입을 좁혀둘까유

aken-you replied:
저는 타이틀에 string만 들어갈 것 같아서 여쭤본 거 였어요!

aken-you commented on src/features/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,36 @@
+import { PasswordChangeModal } from "./_PasswordChangeModal";
+import { settingClassName } from "./setting.styles";
+
+export const PasswordChangeButton = () => {
```
index.ts에서 export 해야해요 👀

aken-you commented on src/pages/my-page/setting/manage-account/page.tsx:
```diff
@@ -0,0 +1,40 @@
+
+export const AccountManagementPage = () => {
+  const navigate = useNavigate();
+  const role = useAuthStore((state) => state.role);
```
useAuthStore의 role 타입이 `string | null`이던데, 더 좁혀보는 건 어떨까요?

현재 role은 null, ROLE_NONE, ROLE_GUEST, ROLE_USER, ROLE_ADMIN 값만 가져요.

yonghyeun replied:
동의합니다 이 부분도 나중에 리팩토링 해야 합니다 

예전 작업에서 `useAuthStore` 의 타입을 좁힐려고 해봤더니 useAuthStore 를 사용하는 다른 컴포넌트 들에서 다 컴파일 에러가 뜨더라고요 

그래서 한 번에 날 잡고 useAuthStore 타입 좁히기 리팩토링 브랜치를 파서 작업 해야겠습니다

aken-you commented on src/entities/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,23 @@
+      <span className="text-grey-700 body-2">{email}</span>
+    </div>
+  );
+};
```
파일 이름을 Account로 하신 이유가 있을까요 ?!

yonghyeun replied:
`settings` 에서도 내부적으로 다양한 도메인들이 존재하더라고요

예를 들어 내 계정 수정 , 계정 정보 , 알림 ... 등등

이 컴포넌트 들을 컴포넌트 하나 하나 쪼개서 `setting` 슬라이스에 넣어두는 것 보다 

내부 도메인마다 하나의 파일에 정의해두는 편이 찾기나 관리 편에서 좋아 보여서요 

동의한다면 현재 `SettingItems` 에서 하나 하나 쪼개져서 만들어져 있는 컴포넌트 들도 이전 방식처럼 `SettingItems` 파일 하나에 집어 넣어둘까 생각중입니다

yonghyeun replied:
네이밍을 왜 이렇게 했냐고 물으신다면 , 계정 관리라서요 , 훔 .. 아니면 AccountManagement 이런 식으로 할까유

aken-you replied:
아 이 파일 안에 AccountEmail만 있는 줄 알았어요~

aken-you commented on src/entities/setting/ui/setting.styles.ts:
```diff
@@ -0,0 +1,2 @@
+export const settingClassName =
+  "py-2 flex justify-between w-full text-grey-700 title-2";
```
settingClassName가 여기저기서 사용돼서 따로 styles 파일에 두신게 맞을까요?

yonghyeun replied:
맞숩니다 `/my-page/setting` 에서 모든 컴포넌트 들이 해당 스타일을 공유 합니다

aken-you commented on src/entities/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,23 @@
+import { settingClassName } from "./setting.styles";
+
+export const AccountEmail = () => {
+  // TODO API 요청이나 userInfo 에서 가져오기
+  const emailFrom: "email" | "naver" | "google" = "email";
+  const email = "gaeun1234@gmail.com";
```
만약에 api 요청으로 email 값을 받을 거라면 여기서 요청할 예정인가요?!

yonghyeun replied:
`userInfo` 를 불러오는 `useQuery` 를 여기서 호출해서 사용하지 않을까 싶습니다 

아니면 뭐 store 에 저장 해둘 수도 있겠고요

aken-you replied:
entities에 있는 컴포넌트라 이 컴포넌트 내부에서 호출해야 하는건지, 아니면 AccountEmail를 사용하는 부모 컴포넌트에서 호출해야하는 건지 고민이 돼서요!

aken-you commented on src/entities/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,23 @@
+
+export const AccountEmail = () => {
+  // TODO API 요청이나 userInfo 에서 가져오기
+  const emailFrom: "email" | "naver" | "google" = "email";
```
이 변수를 emailFrom으로 지은 이유가 있을까요?!

yonghyeun replied:
변수명을 뭐로 짓는지는 너무 고민이 되는 거 같습니다 

뭐가 더 나아보이시나요 

제 의도는 어떤 걸 이용해서 회원 가입했냐를 나타내고 싶었습니다

yonghyeun replied:
`signUpMethod` 로 변경했습니다

aken-you replied:
저도 더 좋은 네이밍이 없을까 고민했는데 어렵더라구요 
확인했습니다!

aken-you commented on src/features/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,36 @@
+    <button className={settingClassName} onClick={handleOpen}>
+      <p>비밀번호 변경</p>
+      <div className="flex">
+        <span className="text-grey-700 body-2">●●●●●●●●</span>
```
이 부분 용현님이 피그마에 코멘트 남긴거 봤어요.
저는 이 표시가 있어야할까? 싶은데 어떻게 생각하시나요?

yonghyeun replied:
저도 개인적으론 자물쇠 아이콘이나 , 아니면 아예 존재하지 않는 편이 나아보이긴 합니다

aken-you replied:
저는 ●●●●●●●● 보단 아예 없는 편이 좋지 않을까 생각하는데, 나중에 수정해야하는게 좋을까요 ?

aken-you commented on src/features/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,36 @@
+      </div>
+    </button>
+  );
+};
```
<img width="751" alt="스크린샷 2024-09-26 오후 7 27 59" src="https://github.com/user-attachments/assets/9e539202-a3da-4cfc-b70e-a64194932aa6">

이 풀리퀘에서 소셜 로그인으로 로그인 한 경우, 렌더링하지 못하도록 처리 못하겠죠...?

yonghyeun replied:
음 일단 하드 코딩 된 값으로 그렇게 해두고 구현도 가능하겠지만 

그렇게 하면 퍼블리싱 단계에서 전체적인 UI 를 테스트 해보기 힘드니 

userInfo 문서가 나오면 그 때 조건부적으로 렌더링 해둘까봐요

aken-you commented on src/pages/my-page/setting/manage-account/page.tsx:
```diff
@@ -0,0 +1,40 @@
+import {
+  AccountCancellation,
+  PasswordChangeButton,
+} from "@/features/setting/ui/Account";
```
여기도 나중에 ui 폴더에서 가져와야 할 것 같아요 👀

yonghyeun replied:
수정했숩니다

aken-you commented on src/features/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,36 @@
+};
+
+// TODO 계정 탈퇴 기능은 기획 완성 후 추가하기
+export const AccountCancellation = () => {
```
Cancellation이 취소라는 의미를 가지고 있던데, 계정 탈퇴 버튼을 `AccountCancellation`로 네이밍을 정한 이유가 무엇인가요?!

yonghyeun replied:
![image](https://github.com/user-attachments/assets/abec11c5-000b-47af-a297-d0641d1ef7cf)
![image](https://github.com/user-attachments/assets/080ba579-df28-47e3-961a-5c139c13731a)

범용적으로 쓰이는 표현인 거 같숩니다

aken-you commented on src/features/setting/ui/_AccountCancellationModal.tsx:
```diff
@@ -0,0 +1,57 @@
+      <CloseNavigationBar onClick={onClose} />
+      <section className="flex flex-col gap-8">
+        {/* 프로덕트 마스코트 */}
+        <img src="/default-image.png" className="mx-auto"></img>
```
단일 태그로 바꿔야 할 것 같아요!

aken-you commented on src/features/setting/ui/_CancellactionCheckModal.tsx:
```diff
@@ -0,0 +1,131 @@
+      {/* 상단 네비게이션 바 */}
+      <div className="flex self-stretch justify-between">
+        <h1 className="text-grey-900 title-1">비밀번호 확인</h1>
+        <button className="px-[0.3125rem]" onClick={handleClose}>
```
급한 건 아니지만 aria-label 달아주시면 좋을 것 같아요!

yonghyeun replied:
작업 완료용

aken-you commented on src/features/setting/ui/_PasswordChangeModal.tsx:
```diff
@@ -0,0 +1,244 @@
+    ? "비밀번호를 입력해 주세요"
+    : isValidNewPassword
+      ? ""
+      : "비밀번호 형식에 맞게 입력해 주세요";
```
현재 비밀번호랑 새 비밀번호가 일치할 경우는 기획서에 빠져있어서 구현하지 않은 게 맞을까요?

yonghyeun replied:
넵 맞숩니다

aken-you commented on src/features/setting/ui/_PasswordChangeModal.tsx:
```diff
@@ -0,0 +1,244 @@
+  };
+
+  // ExitModal 이 Confirm 되면 발생 할 이벤트 핸들러
+  const onExitPasswordChangeModal = () => {
```
이건 컴포넌트 내부에 선언한 핸들러니까 네이밍을 `on*`대신 `handle*`로 하는 건 어떨까요?!

aken-you commented on src/features/setting/ui/_PasswordChangeModal.tsx:
```diff
@@ -0,0 +1,244 @@
+    (state) => state.reset,
+  );
+  // PasswordChangeModal X 아이콘이나 취소 버튼이 클릭 시 호출 되는 이벤트 핸들러
+  const handleClose = () => {
+    const { currentPassword, newPassword, confirmPassword } =
+      usePasswordChangeFormStore.getState();
+    if (currentPassword || newPassword || confirmPassword) {
+      handleOpenExitModal();
+      return;
+    }
+    onClose();
+  };
```
handleOpenExitModal를 사용하고 있으니까 useModal 밑에서 handleClose를 선언하는 건 어떨까요?

aken-you commented on src/features/setting/ui/_CancellactionCheckModal.tsx:
```diff
@@ -0,0 +1,131 @@
+  );
+
+  // CancellationCheckModal X 아이콘이나 취소 버튼이 클릭 시 호출 되는 이벤트 핸들러
+  const handleClose = () => {
+    const { password } = useAccountCancellationFormStore.getState();
+    if (password) {
+      handleOpen();
+      return;
+    }
+    onClose();
+  };
```
handleOpen, onClose를 사용하고 있으니까 useModal 밑에 선언하는 건 어떨까요?

aken-you commented on src/features/setting/ui/_CancellactionCheckModal.tsx:
```diff
@@ -0,0 +1,131 @@
+  };
+
+  // ExitModal 이 Confirm 되면 발생 할 이벤트 핸들러
+  const onExitCancellationModal = () => {
```
컴포넌트 내부에서 선언했으니까 `on*` 대신 `handle*`은 어떨까요?

aken-you commented on src/features/setting/ui/_PasswordChangeModal.tsx:
```diff
@@ -0,0 +1,244 @@
+      </div>
+    </Modal>
+  );
+};
```
어떤 input에 값을 입력하고, 모달 외부를 누르면 confirm modal이 나오지 않습니다.
외부를 눌러서 모달이 닫힌 다음, 다시 비밀번호 변경을 누르면 비밀번호 형식에 맞게 입력하라는 문구가 뜹니다.


https://github.com/user-attachments/assets/2bfec0d2-5985-49c9-9c8d-10ace5715609

yonghyeun replied:
세상에 마상에나 이런 문제가 

꼭 해결해야 하는 문제긴 한데, 이러면 컴포넌트가 매우 지저분해질 거 같긴 합니다 

비밀번호 변경 모달을 트리거 하는 컴포넌트 내부에서 ConfirmModal 을 조건부적으로 불러오는 로직을 짜야 하니까 .. 

씁 고민 좀 해보겠습니다

aken-you commented on src/features/setting/ui/_AccountCancellationModal.tsx:
```diff
@@ -0,0 +1,57 @@
+      </section>
+    </Modal>
+  );
+};
```
버튼 사이즈를 조절해야 할 것 같습니다 👀 

<img width="604" alt="스크린샷 2024-09-26 오후 11 18 24" src="https://github.com/user-attachments/assets/470cce2d-a850-40ab-a2dd-e4d3dbf14225">

yonghyeun replied:
그갹 풀페이지 모달에는 패딩 값이 들어있지 않았군요 

수정했습니다.

(풀페이지 모달을 이미 사용하고 있는 다른 컴포넌트 들에서도 내부에서 `px` 값을 선언해서 쓰고 있어서 풀페이지 모달의 스타일을 수정하지 않고 풀페이지 모달을 호출하고 내부에서 px 값을 사용했습니다.)

> 하면서 느끼는게 디자인 시스템을 빡세게 잘 만들어놔야겠습니다..
> 손 댈 곳이 한 두 개가 아니네요

aken-you commented on src/features/setting/ui/_CancellactionCheckModal.tsx:
```diff
@@ -0,0 +1,131 @@
+          취소
+        </Button>
+      </div>
+    </Modal>
```
여기서도 모달 외부를 누르면 confirm modal이 안뜨네요 👀 


https://github.com/user-attachments/assets/7d7264a4-90b7-4517-9c6d-aa2dbfaff2db

yonghyeun replied:
쓰으으으읍 고민 좀 해보겠습니다 이 부분

aken-you reviewed (commented):
이번에도 수고하셨습니다~

yonghyeun committed:
refactor[#199] : Notice props 타입 좁히기

yonghyeun committed:
refactor[#199] : Account index.ts 공용 API 추가

yonghyeun committed:
refactor[#199] : settingClassName import 하여서 사용하도록 수정

yonghyeun committed:
refactor[#199] : emailFrom -> signUpMethod 로 이름 변경

- 해당 변수가 나타내는 의미를 잘 표현 할 수 있도록 어떤 회원 가입시 어떤 방법으로 가입했는지를 표현

yonghyeun committed:
refactor[#199] : 모달 close 버튼에 aria-label 달기

yonghyeun committed:
fix[#199] : 탈퇴 확인 풀페이지 모달 패딩값 추가

yonghyeun committed:
refactor[#199] : Modal 이벤트 핸들러 선언 순서 변경 & 함수명 변경

yonghyeun committed:
feat[#199] : useOverlay 기능 추가

- beforeClose 로 removeOverlay 를 호출 시키는 것을 reject 시킬 수 있도록 try catch 문 추가
- id 값을 new Date().getTime 이 아닌 랜덤한 값으로 변경

yonghyeun committed:
refactor[#199] : CancellationCheckModal 파일 오타 수정

yonghyeun committed:
feat[#199] : useCancellationCheckModal 생성

yonghyeun committed:
feat[#199] : usePasswordChangeModal 훅 생성

yonghyeun committed:
feat[#199] : 커스텀 훅 이용하여 모달 불러오도록 생성

yonghyeun committed:
fix[#199] : beforeClose 에서 onClose 를 호출하지 않도록 수정

- 재귀 flag 제거

yonghyeun commented:
# 대규모 업데이트가 일어났습니다 

말씀해주신 상황 처럼 `overlay` 밖 부분을 클릭 했을 때 `ConfirmModal` 이 나타나지 않는 경우를 해결했습니다. 

어떻게 해결했는가를 설명하기 앞서 어떤 점이 문제였는지를 먼저 이야기 해보겠습니다. 

# `ConfirmModal` 은 도대체 왜 모달 외부 부분을 클릭 했을 때 나타나지 않았는가 ? 

```tsx
  // 이전 AccountCancellationModal 을 호출 하는 메소드
  const { handleOpen, onClose } = useModal(() => (
    <AccountCancellationModal onClose={onClose} />
  ));
```

만약 오버레이 외부 영역을 클릭 했을 때 호출되는 이벤트 핸들러는 `onClose` 메소드 입니다. 

해당 메소드는 아무런 로직 없이 오버레이 포탈 내부에 존재하는 `AccountCancellationModal ` 컴포넌트를 언마운트 시키기만 합니다. 

그 안에서 `ConfirmModal` 을 조건부적으로 띄우니 마니 하는 비즈니스 로직이 주입 된 상태가 아닙니다. 

# 결과물을 먼저 봅시다

```tsx
export const AccountCancellation = () => {
  const handleOpenCancellationCheckModal = useCancellationCheckModal();

  return (
    <button
      className={settingClassName}
      onClick={handleOpenCancellationCheckModal}
    >
      탈퇴하기
    </button>
  );
```

결과물을 먼저 보게 되면 `AccountCancellationModal` 을 호출하는 쪽에서 특정한 커스텀 훅으로 `handleOpen` 만을 인수로 받습니다. 

그럼 렌더링 되는 `AccountCancellationModal` 내부를 살펴볼까요 ?

```tsx
export const CancellationCheckModal = ({
  onClose,
}: {
  onClose: () => Promise<void>;
}) => {
  return (
    // TODO FormModal 생성 되면 적용하기
    <Modal modalType="center">
```

모달 내부에서 `ExitConfirmModal` 을 호출하는 로직이 제거 되었습니다. 

이제 앞으로 모달들은 컴포넌트 내부에서 본인이 닫히기 전에 컨펌 모달을 띄워야 하는지, 또 어떤 상황에서 띄워야 하는지에 대해 고민하지 않습니다. 

__이제야 말로 컴포넌트가 비즈니스 로직과 분리되어 순수하게 `UI` 에만 집중 할 수 있게 되었습니다 !__

![result good](https://github.com/user-attachments/assets/0ed13758-c072-4fe9-b9c5-d777d52a1d12)

결과물도 매우 잘 작동합니다 짝짝짝 

이제 차례대로 이 문제를 어떻게 해결했는지 적어보겠습니다.

# 1. useOverlay 훅 수정

```tsx
export const useOverlay: UseOverlay = (
  createOverlayComponent,
  options = {},
) => {
  ...
  const id = Math.random(); // 이전엔 new Date().getTime()
  ...

  const onClose = async () => {
    try {
      await beforeClose?.();
      removeOverlay(id);
      setIsOpen(false);
    } catch (error) {
      console.error(error);
    } finally {
      await afterClose?.();
    }
  };
```

`useOverlay` 에서 `id` 값과 `onClose` 부분이 수정 되었습니다. 

`id` 값은 `getTime()` 으로 해놨더니 종종 같은 시간에 생성 된 모달로 인해 유니크한 키 값이 아니라는 경고가 떠 수정했습니다. 

`onClose` 부분의 변화가 가장 큰 부분입니다. try , catch 문을 사용 하여 __`beforeClose` 메소드가 `removeOverlay , setIsOpen` 등의 호출을 중지 시킬 수 있습니다 !!__

`removeOverlay , setIsOpen` 호출을 중지 시켜야 하는 경우는 바로 위 처럼 `ExitConfirmModal` 이 나타나야 하는 경우 입니다. 

![image](https://github.com/user-attachments/assets/f924a58a-896c-4a55-bf6b-8d00f7b9c24b)

예를 덜어 비밀번호 변경 모달의 `onClose` 메소드가 시행 되었을 때 내부의 값이 존재 한다면 `removeOverlays , setIsOpen` 메소드는 호출 되면 안됩니다. 

해당 값을 호출하지 않고 `ExitConfirmModal` 을 띄워야만 합니다. 

# 2. 커스텀 훅 안에서 `Modal` , `ConfirmModal` 을 동시에 생성 하자

```tsx
export const useCancellationCheckModal = () => {
  const resetCancellationForm = useAccountCancellationFormStore(
    (state) => state.reset,
  );

  const {
    handleOpen: handleOpenExitConfirmModal,
    onClose: onCloseConfirmModal,
  } = useModal(() => (
    <ExitConfirmModal
      onConfirm={() => {
        resetCancellationForm();
        onCloseConfirmModal();
        onCloseCancellationCheckModal();
      }}
      onClose={onCloseConfirmModal}
    />
  ));

  const {
    handleOpen: handleOpenCancellationCheckModal,
    onClose: onCloseCancellationCheckModal,
  } = useModal(
    () => <CancellationCheckModal onClose={onCloseCancellationCheckModal} />,
    {
      /* 이 부분은 추후 설명 */
    },
  );
  return handleOpenCancellationCheckModal;
};
```

해당 커스텀 훅은 회원 탈퇴 모달을 여는 커스텀 훅입니다. 

 예전과 다른 점은 모달 안에서 다른 모달을 호출했던 예전 방식과 다르게 커스텀 훅 안에서 열고자 하는 모달과 열고자 하는 모달에서 열고자 하는 모달(여기선 `ExitConfirmModal`) 이 동시에 선언 된다는 점입니다. (이하 모달을 트리거 하는 컴포넌트를 `Trigger` , `Trigger`가 열고자 하는 모달을 `Modal` , `Modal` 에서 열고자 하는 모달을 `ConfirmModal` 이라 칭하겠습니다.)

> 이런 모달 생성 순서 변화로 인해 `overlay` 에서 `id` 값을 `getTime` 이 아닌 난수로 수정해주었습니다. 

이렇게 하게 되면 우선 모달의 재사용성이 매우 올라갑니다 !

모달은 이제 우선 순수하게 본인을 닫는 `onClose` 만 외부에서 `props`로 받아 부착해주면 됩니다. 더 이상 어떤 `ExitConfirmModal` 을 띄워야 할지 컴포넌트 안에서 호출 할 필요가 없습니다.

이렇게 `Modal` 과 `ConfirmModal` 들의 결합을 매우 느슨하게 만들어줬습니다. 

이로 인해 어떤 모달이 `ConfirmModal` 이나 또 다른 모달 들을 띄워줘야 할 때 내부에서 어떤 모달을 `useModal` 로 호출해 생성하는 것이 아니기에 모달의 재사용성이 매우 올라갑니다. 

# 3. `Modal` 과 `ConfirmModal` 을 이어주자 

예전 `useOverlay` 훅을 만들 때 애니메이션 효과를 넣기 위해 넣어줬던 `beforeClose , afterClose` 메소드를 이용해 `Modal` 과 `ConfirmModal`을 이어줬습니다. 

```tsx
export const useChangePasswordModal = () => {
...
const {
    handleOpen: handleOpenPasswordChangeModal,
    onClose: onClosePasswordChangeModal,
  } = useModal(
    () => <PasswordChangeModal onClose={onClosePasswordChangeModal} />,
    {
      beforeClose: () => {
        if (!beforeCloseRecursiveFlag) {
          return;
        }

        const { currentPassword, newPassword, confirmPassword } =
          usePasswordChangeFormStore.getState();

        if (currentPassword || newPassword || confirmPassword) {
          handleOpenExitConfirmModal();
          throw new Error("입력 중인 폼이 존재하여 ConfirmModal을 열었습니다.");
        }
      },
    },
  );
```

예를 들어 `PasswordChangeModal ` 를 생성하는 `useModal` 을 살펴 봅시다. 

`beforeClose` 에서 조건부적으로 `ConfirmModal` 을 띄우고 에러를 던져 `PasswordCangeModal` 가 닫히는 행위를 중지 시킵니다. 
> `useOverlay` 에서 `try ,catch` 를 이용해  에러가 발생하면 `removeOverlay , setIsOpen` 를 호출하지 않던 경우를 기억해보세요 ! 

이렇게 `beforeClose` 에서 `ConfirmModal` 을 띄울지 말지에 대한 로직을 넣어줌으로서 앞으로 모달 외부 영역을 클릭 했을 때 `OverlayPortal` 에서 호출되는 `onClose` 메소드는 우선적으로 `beforeClose` 메소드를 먼저 시행 한 후 해당 모달을 닫을지 말지를 결정 합니다. 

---

# 정리 

이번 작업에선 다음과 같은 일들이 일어났습니다.

1. `useOverlay` 에서 `beforeClose` -> `close` ->`afterClose`  단계에서  `beforeClose` 가 `close` 의 호출을 중지 시킬 수 있습니다. 
2. 각 모달 들은 어떤 `ConfirmModal` 을 띄워야 할지 결정하지 않고 말 그대로 본인을 닫는 `onClose` 메소드 만을 신경 쓰게 됩니다. 
3. `ConfirmModal` 을 띄워야 하는 모달의 경우엔 `useModal` 로 생성 될 때 `beforeClose` 에서 조건부적으로 `ConfirmModal` 을 띄울지 말지를 선언 하고 에러를 던져 `close` 단계를 중지 시킵니다.
4. 3번 단계를 커스텀 훅으로 컴포넌트 밖으로 빼내어 `Modal` 을 호출하는 컴포넌트에선 순수하게 본인의 렌더링 결과와 모달을 여는 트리거인 `handleOpen` 에만 신경 쓸 수 있게 되었습니다.

yonghyeun commented:
# 대규모 업데이트가 일어났습니다 

말씀해주신 상황 처럼 `overlay` 밖 부분을 클릭 했을 때 `ConfirmModal` 이 나타나지 않는 경우를 해결했습니다. 

어떻게 해결했는가를 설명하기 앞서 어떤 점이 문제였는지를 먼저 이야기 해보겠습니다. 

# `ConfirmModal` 은 도대체 왜 모달 외부 부분을 클릭 했을 때 나타나지 않았는가 ? 

```tsx
  // 이전 AccountCancellationModal 을 호출 하는 메소드
  const { handleOpen, onClose } = useModal(() => (
    <AccountCancellationModal onClose={onClose} />
  ));
```

만약 오버레이 외부 영역을 클릭 했을 때 호출되는 이벤트 핸들러는 `onClose` 메소드 입니다. 

해당 메소드는 아무런 로직 없이 오버레이 포탈 내부에 존재하는 `AccountCancellationModal ` 컴포넌트를 언마운트 시키기만 합니다. 

그 안에서 `ConfirmModal` 을 조건부적으로 띄우니 마니 하는 비즈니스 로직이 주입 된 상태가 아닙니다. 

`ConfirmModal` 을 띄울지 말지는 `AccountCancellationModal ` 내부에서 `X` 버튼을 누르거나 취소 버튼을 눌렀을 때에만 결정됩니다.

# 결과물을 먼저 봅시다

```tsx
export const AccountCancellation = () => {
  const handleOpenCancellationCheckModal = useCancellationCheckModal();

  return (
    <button
      className={settingClassName}
      onClick={handleOpenCancellationCheckModal}
    >
      탈퇴하기
    </button>
  );
```

결과물을 먼저 보게 되면 `AccountCancellationModal` 을 호출하는 쪽에서 특정한 커스텀 훅으로 `handleOpen` 만을 인수로 받습니다. 

그럼 렌더링 되는 `AccountCancellationModal` 내부를 살펴볼까요 ?

```tsx
export const CancellationCheckModal = ({
  onClose,
}: {
  onClose: () => Promise<void>;
}) => {
  return (
    // TODO FormModal 생성 되면 적용하기
    <Modal modalType="center">
```

모달 내부에서 `ExitConfirmModal` 을 호출하는 로직이 제거 되었습니다. 

이제 앞으로 모달들은 컴포넌트 내부에서 본인이 닫히기 전에 컨펌 모달을 띄워야 하는지, 또 어떤 상황에서 띄워야 하는지에 대해 고민하지 않습니다. 

__이제야 말로 컴포넌트가 비즈니스 로직과 분리되어 순수하게 `UI` 에만 집중 할 수 있게 되었습니다 !__

![result good](https://github.com/user-attachments/assets/0ed13758-c072-4fe9-b9c5-d777d52a1d12)

결과물도 매우 잘 작동합니다 짝짝짝 

이제 차례대로 이 문제를 어떻게 해결했는지 적어보겠습니다.

# 1. `Overlay` 의 `beforeClose` 변경 

```tsx
// shared/overlay/store/overlay.ts
type StopCloseFlag = void | boolean | null;

export interface OverlayOptions {
  disableInteraction?: boolean;
  // beforeClose 는 close 단계를 시행 할지 말지를 나타내는 StopCloseFlag 를 반환합니다.
  // 만약 StopCloseFlag 가 true 라면  close 단계를 시행하지 않습니다.
  beforeClose?: () => StopCloseFlag | Promise<StopCloseFlag>;
  afterClose?: () => void | Promise<void>;
}
```

이전엔 `beforeClose` 가 아무런 반환값도 뱉지 않았지만 이젠 `StopCloseFlag` 를 반환 하도록 합니다. 

`StopCloseFlag` 는 말 그대로 `beforeClose` 이후에 `close` 단계를 진행 할 것인지를 나타내는 값입니다. 

# 2. `useOverlay` 에서 정의되는 `onClose` 수정 


```tsx
export const useOverlay: UseOverlay = (
  createOverlayComponent,
  options = {},
) => {

  const id = Math.random();

  const onClose = async () => {
    const stopCloseFlag = await beforeClose?.();
    if (stopCloseFlag) {
      return;
    }
    removeOverlay(id);
    setIsOpen(false);
    await afterClose?.();
  };
```

`useOverlay` 에서 `id` 값과 `onClose` 부분이 수정 되었습니다. 

`id` 값은 `getTime()` 으로 해놨더니 종종 같은 시간에 생성 된 모달로 인해 유니크한 키 값이 아니라는 경고가 떠 수정했습니다. 

`onClose` 부분의 변화가 가장 큰 부분입니다.  `beforeClose` 메소드가 `removeOverlay , setIsOpen` 등의 호출을 중지 시킬 수 있습니다 !!__

`beforeClose` 가 만약 `true` 값을 반환한다면 `close` 단계는 진행 되지 않습니다.

`removeOverlay , setIsOpen` 호출을 중지 시켜야 하는 경우는 바로 위 처럼 `ExitConfirmModal` 이 나타나야 하는 경우 입니다. 

![image](https://github.com/user-attachments/assets/f924a58a-896c-4a55-bf6b-8d00f7b9c24b)

예를 덜어 비밀번호 변경 모달의 `onClose` 메소드가 시행 되었을 때 내부의 값이 존재 한다면 `removeOverlays , setIsOpen` 메소드는 호출 되면 안됩니다. 

해당 값을 호출하지 않고 `ExitConfirmModal` 을 띄워야만 합니다. 

# 2. 커스텀 훅 안에서 `Modal` , `ConfirmModal` 을 동시에 생성 하자

```tsx
export const useCancellationCheckModal = () => {
  const resetCancellationForm = useAccountCancellationFormStore(
    (state) => state.reset,
  );

  const {
    handleOpen: handleOpenExitConfirmModal,
    onClose: onCloseConfirmModal,
  } = useModal(() => (
    <ExitConfirmModal
      onConfirm={() => {
        resetCancellationForm();
        onCloseConfirmModal();
        onCloseCancellationCheckModal();
      }}
      onClose={onCloseConfirmModal}
    />
  ));

  const {
    handleOpen: handleOpenCancellationCheckModal,
    onClose: onCloseCancellationCheckModal,
  } = useModal(
    () => <CancellationCheckModal onClose={onCloseCancellationCheckModal} />,
    {
      beforeClose: () => {
         /* 이 부분은 추후 설명 */
        }
      },
    },
  );
  return handleOpenCancellationCheckModal;
};
```

해당 커스텀 훅은 회원 탈퇴 모달을 여는 커스텀 훅입니다. 

 예전과 다른 점은 모달 안에서 다른 모달을 호출했던 예전 방식과 다르게 커스텀 훅 안에서 열고자 하는 모달과 열고자 하는 모달에서 열고자 하는 모달(여기선 `ExitConfirmModal`) 이 동시에 선언 된다는 점입니다. (이하 모달을 트리거 하는 컴포넌트를 `Trigger` , `Trigger`가 열고자 하는 모달을 `Modal` , `Modal` 에서 열고자 하는 모달을 `ConfirmModal` 이라 칭하겠습니다.)

> 이런 모달 생성 순서 변화로 인해 `overlay` 에서 `id` 값을 `getTime` 이 아닌 난수로 수정해주었습니다. 

이렇게 하게 되면 우선 모달의 재사용성이 매우 올라갑니다 !

모달은 이제 우선 순수하게 본인을 닫는 `onClose` 만 외부에서 `props`로 받아 부착해주면 됩니다. 더 이상 어떤 `ExitConfirmModal` 을 띄워야 할지 컴포넌트 안에서 호출 할 필요가 없습니다.

이렇게 `Modal` 과 `ConfirmModal` 들의 결합을 매우 느슨하게 만들어줬습니다. 

이로 인해 어떤 모달이 `ConfirmModal` 이나 또 다른 모달 들을 띄워줘야 할 때 내부에서 어떤 모달을 `useModal` 로 호출해 생성하는 것이 아니기에 모달의 재사용성이 증가 합니다.

# 3. `Modal` 과 `ConfirmModal` 을 이어주자 

예전 `useOverlay` 훅을 만들 때 애니메이션 효과를 넣기 위해 넣어줬던 `beforeClose , afterClose` 메소드를 이용해 `Modal` 과 `ConfirmModal`을 이어줬습니다. 

```tsx
export const useCancellationCheckModal = () => {

  const {
    handleOpen: handleOpenExitConfirmModal,
    onClose: onCloseConfirmModal,
  } = useModal(() => (
    <ExitConfirmModal
     ...
    />
  ));

  ...
  const {
    handleOpen: handleOpenCancellationCheckModal,
    onClose: onCloseCancellationCheckModal,
  } = useModal(
    () => <CancellationCheckModal onClose={onCloseCancellationCheckModal} />,
    {
      beforeClose: () => {
        const { password } = useAccountCancellationFormStore.getState();
        if (password) {
          handleOpenExitConfirmModal();
          return true;
        }
      },
    },
  );
  return handleOpenCancellationCheckModal;
```

예를 들어 `CancellationCheckModal` 를 생성하는 `useModal` 을 살펴 봅시다. 

`beforeClose` 에서 `CancellationCheckModal ` 내부에서 사용하는 `form` 에서 값이 존재 한다면 `ExitConfirmModal` 을 띄우고 `StopCloseFlag` 값을 `true` 로 반환하여 `close` 단계를 진행하지 않습니다.

> `useOverlay` 에서  `beforeClose` 가 반환하는 값이 `true` 라면 removeOverlay , setIsOpen` 를 호출하지 않던 경우를 기억해보세요 ! 

# `beforeClose` 를 추가해준 것이 왜 이 기능을 동작하게 만들었는가 ? 

문제의 원인을 다시 되짚어봐야 합니다. 

```tsx
  // 이전 AccountCancellationModal 을 호출 하는 메소드
  const { handleOpen, onClose } = useModal(() => (
    <AccountCancellationModal onClose={onClose} />
  ));
```

이전 작업물에서 `useModal` 훅이 반환하는 `onClose` 메소드는 `useModal` 호출과 함께 생성되어 `OverlayPortal - OverlayWrapper` 로 넘어가 모달 외부 영역을 클릭 시 호출되는 메소드였습니다. 

이전 문제에서 `OverlayWrapper` 에서 호출 되는 `onClose` 는 `ExitConfirmModal` 을 띄울지 말지에 대한 처리가 존재하지 않습니다. 

그래서 외부 영역을 클릭 해도 컨펌 모달이 뜨지 않았던 거죠 

```tsx
export const useCancellationModal () => {
....
  const {
    handleOpen: handleOpenCancellationCheckModal,
    onClose: onCloseCancellationCheckModal,
  } = useModal(
    () => <CancellationCheckModal onClose={onCloseCancellationCheckModal} />,
    {
      beforeClose: () => {
        const { password } = useAccountCancellationFormStore.getState();
        if (password) {
          handleOpenExitConfirmModal();
          return true;
        }
      },
    },
  );
  return handleOpenCancellationCheckModal;
};
...
}
```

하지만 수정 된 지금은 다릅니다. `OverlayWrapper` 에서 호출 되는 `onClose` 메소드는 `useModal` 호출 시 인수로 건내준 `beforeClose` 를 시행 후 `clsoe -> afterClose` 단계를 거칩니다. 

이제 `OverlayWrapper` 에서도 `ExitConfirmModal` 을 띄울지 말지에 대한 로직이 존재하여 정상 작동 하게 됩니다 :) 

---

# 정리 

이번 작업에선 다음과 같은 일들이 일어났습니다.

1. `useOverlay` 에서 `beforeClose` -> `close` ->`afterClose`  단계에서  `beforeClose` 가 `close` 의 호출을 중지 시킬 수 있습니다. 
2. 각 모달 들은 어떤 `ConfirmModal` 을 띄워야 할지 결정하지 않고 말 그대로 본인을 닫는 `onClose` 메소드 만을 신경 쓰게 됩니다. 
3. `ConfirmModal` 을 띄워야 하는 모달의 경우엔 `useModal` 로 생성 될 때 `beforeClose` 에서 조건부적으로 `ConfirmModal` 을 띄울지 말지를 결정하고 `StopCloseFlag`  를 반환 합니다. 만약 `true` 값이 반환 된다면 `close` 단계는 시행 되지 않습니다.
4. 3번 단계를 커스텀 훅으로 컴포넌트 밖으로 빼내어 `Modal` 을 호출하는 컴포넌트에선 순수하게 본인의 렌더링 결과와 모달을 여는 트리거인 `handleOpen` 에만 신경 쓸 수 있게 되었습니다.

yonghyeun commented:
# 대규모 업데이트가 일어났습니다 

말씀해주신 상황 처럼 `overlay` 밖 부분을 클릭 했을 때 `ConfirmModal` 이 나타나지 않는 경우를 해결했습니다. 

어떻게 해결했는가를 설명하기 앞서 어떤 점이 문제였는지를 먼저 이야기 해보겠습니다. 

# `ConfirmModal` 은 도대체 왜 모달 외부 부분을 클릭 했을 때 나타나지 않았는가 ? 

```tsx
  // 이전 AccountCancellationModal 을 호출 하는 메소드
  const { handleOpen, onClose } = useModal(() => (
    <AccountCancellationModal onClose={onClose} />
  ));
```

만약 오버레이 외부 영역을 클릭 했을 때 호출되는 이벤트 핸들러는 `onClose` 메소드 입니다. 

해당 메소드는 아무런 로직 없이 오버레이 포탈 내부에 존재하는 `AccountCancellationModal ` 컴포넌트를 언마운트 시키기만 합니다. 

그 안에서 `ConfirmModal` 을 조건부적으로 띄우니 마니 하는 비즈니스 로직이 주입 된 상태가 아닙니다. 

# 결과물을 먼저 봅시다

```tsx
export const AccountCancellation = () => {
  const handleOpenCancellationCheckModal = useCancellationCheckModal();

  return (
    <button
      className={settingClassName}
      onClick={handleOpenCancellationCheckModal}
    >
      탈퇴하기
    </button>
  );
```

결과물을 먼저 보게 되면 `AccountCancellationModal` 을 호출하는 쪽에서 특정한 커스텀 훅으로 `handleOpen` 만을 인수로 받습니다. 

그럼 렌더링 되는 `AccountCancellationModal` 내부를 살펴볼까요 ?

```tsx
export const CancellationCheckModal = ({
  onClose,
}: {
  onClose: () => Promise<void>;
}) => {
  return (
    // TODO FormModal 생성 되면 적용하기
    <Modal modalType="center">
```

모달 내부에서 `ExitConfirmModal` 을 호출하는 로직이 제거 되었습니다. 

이제 앞으로 모달들은 컴포넌트 내부에서 본인이 닫히기 전에 컨펌 모달을 띄워야 하는지, 또 어떤 상황에서 띄워야 하는지에 대해 고민하지 않습니다. 

__이제야 말로 컴포넌트가 비즈니스 로직과 분리되어 순수하게 `UI` 에만 집중 할 수 있게 되었습니다 !__

![result good](https://github.com/user-attachments/assets/0ed13758-c072-4fe9-b9c5-d777d52a1d12)

결과물도 매우 잘 작동합니다 짝짝짝 

이제 차례대로 이 문제를 어떻게 해결했는지 적어보겠습니다.

# 1. useOverlay 훅 수정

```tsx
export const useOverlay: UseOverlay = (
  createOverlayComponent,
  options = {},
) => {
  ...
  const id = Math.random(); // 이전엔 new Date().getTime()
  ...

  const onClose = async () => {
    try {
      await beforeClose?.();
      removeOverlay(id);
      setIsOpen(false);
    } catch (error) {
      console.error(error);
    } finally {
      await afterClose?.();
    }
  };
```

`useOverlay` 에서 `id` 값과 `onClose` 부분이 수정 되었습니다. 

`id` 값은 `getTime()` 으로 해놨더니 종종 같은 시간에 생성 된 모달로 인해 유니크한 키 값이 아니라는 경고가 떠 수정했습니다. 

`onClose` 부분의 변화가 가장 큰 부분입니다. try , catch 문을 사용 하여 __`beforeClose` 메소드가 `removeOverlay , setIsOpen` 등의 호출을 중지 시킬 수 있습니다 !!__

`removeOverlay , setIsOpen` 호출을 중지 시켜야 하는 경우는 바로 위 처럼 `ExitConfirmModal` 이 나타나야 하는 경우 입니다. 

![image](https://github.com/user-attachments/assets/f924a58a-896c-4a55-bf6b-8d00f7b9c24b)

예를 덜어 비밀번호 변경 모달의 `onClose` 메소드가 시행 되었을 때 내부의 값이 존재 한다면 `removeOverlays , setIsOpen` 메소드는 호출 되면 안됩니다. 

해당 값을 호출하지 않고 `ExitConfirmModal` 을 띄워야만 합니다. 

# 2. 커스텀 훅 안에서 `Modal` , `ConfirmModal` 을 동시에 생성 하자

```tsx
export const useCancellationCheckModal = () => {
  const resetCancellationForm = useAccountCancellationFormStore(
    (state) => state.reset,
  );

  const {
    handleOpen: handleOpenExitConfirmModal,
    onClose: onCloseConfirmModal,
  } = useModal(() => (
    <ExitConfirmModal
      onConfirm={() => {
        resetCancellationForm();
        onCloseConfirmModal();
        onCloseCancellationCheckModal();
      }}
      onClose={onCloseConfirmModal}
    />
  ));

  const {
    handleOpen: handleOpenCancellationCheckModal,
    onClose: onCloseCancellationCheckModal,
  } = useModal(
    () => <CancellationCheckModal onClose={onCloseCancellationCheckModal} />,
    {
      /* 이 부분은 추후 설명 */
    },
  );
  return handleOpenCancellationCheckModal;
};
```

해당 커스텀 훅은 회원 탈퇴 모달을 여는 커스텀 훅입니다. 

 예전과 다른 점은 모달 안에서 다른 모달을 호출했던 예전 방식과 다르게 커스텀 훅 안에서 열고자 하는 모달과 열고자 하는 모달에서 열고자 하는 모달(여기선 `ExitConfirmModal`) 이 동시에 선언 된다는 점입니다. (이하 모달을 트리거 하는 컴포넌트를 `Trigger` , `Trigger`가 열고자 하는 모달을 `Modal` , `Modal` 에서 열고자 하는 모달을 `ConfirmModal` 이라 칭하겠습니다.)

> 이런 모달 생성 순서 변화로 인해 `overlay` 에서 `id` 값을 `getTime` 이 아닌 난수로 수정해주었습니다. 

이렇게 하게 되면 우선 모달의 재사용성이 매우 올라갑니다 !

모달은 이제 우선 순수하게 본인을 닫는 `onClose` 만 외부에서 `props`로 받아 부착해주면 됩니다. 더 이상 어떤 `ExitConfirmModal` 을 띄워야 할지 컴포넌트 안에서 호출 할 필요가 없습니다.

이렇게 `Modal` 과 `ConfirmModal` 들의 결합을 매우 느슨하게 만들어줬습니다. 

이로 인해 어떤 모달이 `ConfirmModal` 이나 또 다른 모달 들을 띄워줘야 할 때 내부에서 어떤 모달을 `useModal` 로 호출해 생성하는 것이 아니기에 모달의 재사용성이 매우 올라갑니다. 

# 3. `Modal` 과 `ConfirmModal` 을 이어주자 

예전 `useOverlay` 훅을 만들 때 애니메이션 효과를 넣기 위해 넣어줬던 `beforeClose , afterClose` 메소드를 이용해 `Modal` 과 `ConfirmModal`을 이어줬습니다. 

```tsx
export const useChangePasswordModal = () => {
...
const {
    handleOpen: handleOpenPasswordChangeModal,
    onClose: onClosePasswordChangeModal,
  } = useModal(
    () => <PasswordChangeModal onClose={onClosePasswordChangeModal} />,
    {
      beforeClose: () => {
        if (!beforeCloseRecursiveFlag) {
          return;
        }

        const { currentPassword, newPassword, confirmPassword } =
          usePasswordChangeFormStore.getState();

        if (currentPassword || newPassword || confirmPassword) {
          handleOpenExitConfirmModal();
          throw new Error("입력 중인 폼이 존재하여 ConfirmModal을 열었습니다.");
        }
      },
    },
  );
```

예를 들어 `PasswordChangeModal ` 를 생성하는 `useModal` 을 살펴 봅시다. 

`beforeClose` 에서 조건부적으로 `ConfirmModal` 을 띄우고 에러를 던져 `PasswordCangeModal` 가 닫히는 행위를 중지 시킵니다. 
> `useOverlay` 에서 `try ,catch` 를 이용해  에러가 발생하면 `removeOverlay , setIsOpen` 를 호출하지 않던 경우를 기억해보세요 ! 

이렇게 `beforeClose` 에서 `ConfirmModal` 을 띄울지 말지에 대한 로직을 넣어줌으로서 앞으로 모달 외부 영역을 클릭 했을 때 `OverlayPortal` 에서 호출되는 `onClose` 메소드는 우선적으로 `beforeClose` 메소드를 먼저 시행 한 후 해당 모달을 닫을지 말지를 결정 합니다. 

---

# 정리 

이번 작업에선 다음과 같은 일들이 일어났습니다.

1. `useOverlay` 에서 `beforeClose` -> `close` ->`afterClose`  단계에서  `beforeClose` 가 `close` 의 호출을 중지 시킬 수 있습니다. 
2. 각 모달 들은 어떤 `ConfirmModal` 을 띄워야 할지 결정하지 않고 말 그대로 본인을 닫는 `onClose` 메소드 만을 신경 쓰게 됩니다. 
3. `ConfirmModal` 을 띄워야 하는 모달의 경우엔 `useModal` 로 생성 될 때 `beforeClose` 에서 조건부적으로 `ConfirmModal` 을 띄울지 말지를 선언 하고 에러를 던져 `close` 단계를 중지 시킵니다.
4. 3번 단계를 커스텀 훅으로 컴포넌트 밖으로 빼내어 `Modal` 을 호출하는 컴포넌트에선 순수하게 본인의 렌더링 결과와 모달을 여는 트리거인 `handleOpen` 에만 신경 쓸 수 있게 되었습니다.

yonghyeun commented:
> ... 외부를 눌러서 모달이 닫힌 다음, 다시 비밀번호 변경을 누르면 비밀번호 형식에 맞게 입력하라는 문구가 뜹니다.
> 
>  2024-09-26.11.14.16.mov

아 ! 추가로 이 부분도 해결이 됐습니다. 

예전에 저런 문제가 왜 발생했냐면 외부 영역을 클릭하여 모달이 닫혔을 때 내부에 존재하는 `use .. store` 가 reset 되지 않았기 때문입니다 .

> 우리 눈에 보이진 않지만 다시 마운트 되었을 때의 `store` 는 닫히기 전 입력했던 값이 저장 된 상태들이였겠네요 

예전엔 `reset` 메소드가 호출되는 경우는 `ConfirmModal` 에서 확인 버튼을 눌렀을 때에만 `reset` 이 호출되었습니다. 

이젠 외부 영역을 클릭해도 `ConfirmModal` 이 뜨기 때문에 해당 문제는 발생하지 않습니다 !

yonghyeun committed:
feat[#199] : try catch 문을 제거하고 StopCloseFlag 를 이용해 early return 하여 close 단계를 탈출

yonghyeun committed:
refactor[#199] : DividerLine 적용

yonghyeun committed:
refactor[#199] : useOverlay 내부 flag -> stopCloseFlag 로 이름 명확하게 변경

aken-you commented on src/features/setting/ui/Account.tsx:
```diff
@@ -0,0 +1,36 @@
+import { ArrowRightIcon } from "@/shared/ui/icon";
+import { useCancellationCheckModal, useChangePasswordModal } from "../lib";
+import { settingClassName } from "./setting.styles";
```
features/setting에 있는 settingClassName이랑 entities/setting에 있는 settingClassName이랑 거의 동일하지 않나요?!

yonghyeun replied:
완전 동일 합니다 

그런데 그렇다고 해당 스타일을 shared 에 넣기엔 도메인에 종속되는 스타일이라서요

aken-you commented on src/shared/lib/overlay.ts:
```diff
@@ -16,13 +16,16 @@ export const useOverlay: UseOverlay = (
   const [isOpen, setIsOpen] = useState<boolean>(false);
   const { disableInteraction = true, beforeClose, afterClose } = options;
-  const id = new Date().getTime(); // 커스텀 훅이 호출 될 때 마다 유일한 id를 생성합니다.
+  const id = Math.random();
```
Math.random은 확률적으로 중복된 값이 나올 수 있을 것 같은데, 다른 방법으로 하는 건 어떨 것 같나요?!

yonghyeun committed:
fix[#199] : AccountCancellation 컴포넌트에 적절한 Modal 붙이기

yonghyeun committed:
refactor[#199] : AccountCancellation -> AccountCancellationButton 으로 이름 변경

yonghyeun committed:
refactor[#199] : PasswordChangeButton 에서 비밀번호  제거

yonghyeun committed:
fix[#199] : id 값은 Math.random 이 아닌 crypto 를 이용하여 암호화
